### DISCUSSION TITLE
Elements per second in progress bars

### DISCUSSION THREAD

**Discussion by Poshi** [Textualize/rich#2035] [Ideas]
I also realized that you can easily show transfer speed and file size over total file size information in progress bars, but I cannot find columns for the generalized case: elements over total elements and elements per second (speed).
It would make sense to have these columns readily available for general use.
I see that files could be the most common elements being tracked, but "elements" in general will be quite useful too.

**Comment by willmcgugan** [comment]
Does the second paragraph here help? 

https://rich.readthedocs.io/en/latest/progress.html#columns

**Reply by jfthuong** [reply]
Nothing that number over total is `[MofNCompleteColumn](https://rich.readthedocs.io/en/latest/reference/progress.html#rich.progress.MofNCompleteColumn)` ... but it's not very clear how to do " number of elements per seconds", with the possibility to set the unit of such elements (e.g. "files per second", "tests per second")...

**Comment by Poshi** [comment]
It helps, but it doesn't fully cover.
Yes, that could solve one of issues at hand. But then I wonder why there's a DonwloadColumn class, which is the same but adding the units.
But the other issue, getting the speed in elements per second, is not easily solved, as this involved some time tracking.
Sorry if I'm proposing something that can already be easily done, but I've been unable to find it :-/

**Comment by bsridatta** [comment]
Though its possible to customize the progress bar and use MofNColumns, it would be great to have this as an option is basic `track` method. 

For example tqdm, shows M/N and also shows the rate at which the elements are progressing. In all my use cases, though having % of completion is nice, I want to see M/N and rate of progress. Number of files crunches, time to crunch a file. If this is out of box with `track` or with a simple parameter like `format="m/n"` would make it easy to move to rich. Also you dont need to do progress.print allowing to use it in a one-liner like `progress.track`.

If I only need progress bar I feel its easy to just wrap with tqdm instead of defining a progress bar in all my scripts.

I would love to contribute to this if you see the merit in this.

**Reply by willmcgugan** [reply]
If you like tqdm, you could use the integrated Rich support. But I'm unlikely to accept a change that adds parameter(s) for one user's use case. Especially when it can already be done in just a few lines.

If you have a more concrete proposal or a PR I would consider it.

**Reply by bsridatta** [reply]
I dont believe its just me as others are trying to do the same. But cool, thanks for the response.

**Comment by fschoenm** [comment]
You can write your own column by combining some existing code to achieve this:

```python
from rich.progress import ProgressColumn, Task, filesize
from rich.text import Text 

class RateColumn(ProgressColumn):
    """Renders human readable processing rate."""

    def render(self, task: "Task") -> Text:
        """Render the speed in iterations per second."""
        speed = task.finished_speed or task.speed
        if speed is None:
            return Text("", style="progress.percentage")
        unit, suffix = filesize.pick_unit_and_suffix(
            int(speed),
            ["", "×10³", "×10⁶", "×10⁹", "×10¹²"],
            1000,
        )
        data_speed = speed / unit
        return Text(f"{data_speed:.1f}{suffix} it/s", style="progress.percentage")
```
